### DISCUSSION TITLE
Can you read multiple DS18B20 probes with Tasmota on a Sonoff THR320D?

### DISCUSSION THREAD

**Discussion by wayner9** [arendst/Tasmota#19937] [Q&A] [answered]
I flashed a THR320D with Tasmota and set the GPIO25 to DS18x20 to use with the Sonoff temperature probe.  I can see the temp displayed on the screen and I can read it from an API call as in "http://192.168.1.123/cm?cmnd=Status%208"
```
{
    "StatusSNS": {
        "Time": "2023-11-05T12:49:05",
        "DS18B20": {
            "Id": "3C530457A8AC",
            "Temperature": 23.7
        },
        "TempUnit": "C"
    }
}
```

I am also building my own longer probe using a generic waterproof DS18B20 from Amazon.  These cable are dirt cheap.

One question - if you had multiple temperature probes connected to the GPIO25 would you be able to read these with Tasmota?  I know that the 1Wire bus supports multiple cables, and you can do this with a Raspberry Pi.  Would these just show up as multiple probe Ids when calling Status 8?

**Comment by barbudor** [comment, marked as answer]
Tasmota support by default up to 8 DS on the same GPIO. More are possible with self compile.

They indeed would be listed in the SENSOR message either through the Status8 command or better through MQTT.
You have options that requires self compile to use DSAlias command to give a name to each sensor by it's 1-wire ID

**Reply by sfromis** [reply]
In case of trouble, I'd try adding extra pullup when using multiple DS18B20 sensors. IDK how much the THR320D already has.

**Reply by wayner9** [reply]
Thanks.  I will be using no more than 2-3 sensors.  I will likely be using HTTP rather than MQTT.  Right now I am just monitoring and want to output the readings to a CSV file or time series database.

**Reply by wayner9** [reply]
FYI, I have two DS18B20 sensors on one device and here is the JSON response from a Status 8 Command - note that you get a DS18B20-1 and a DS18B20-2 :
```
{
    "StatusSNS": {
        "Time": "2023-11-11T17:36:25",
        "DS18B20-1": {
            "Id": "B77A201864FF",
            "Temperature": 39.2
        },
        "DS18B20-2": {
            "Id": "7B79201864FF",
            "Temperature": 13.1
        },
        "TempUnit": "C"
    }
}
```

**Reply by wayner9** [reply]
One more thing on this - the screen on the THR320D shows both temperature readings.  Anyone knows what is displayed on the screen if you have more than two probes?  Does it only show the first two?  Does it cycle through them two at a time?

**Reply by sfromis** [reply]
Not tried it, but IIRC it will rotate them, where you can choose `DspSpeed`.

**Comment by wayner9** [comment]
I don't know if this is the cheap cables or another issues, but I am seeing periodic spikes that are bad readings.  I guess I will have to filter for these.   The spikes occur on both probes at the same time as you can see.

![ds18b20spikes](https://github.com/arendst/Tasmota/assets/10366145/ddb3b78e-ddf9-4c0d-a100-2fd212b31714)

**Reply by sfromis** [reply]
The synchronicity of the spikes might indicate some sort of electrical noise picked up. I'm wondering if some appliance is switching on or off at those times. While tub heating does not look like the only factor, one of the spikes does coincide. Is there some external factor causing the tub temperature to spike down before heating cycles, or can the suspect spikes be both down and up?

Seeing raw data from Tasmota logs might be useful.

**Reply by wayner9** [reply]
I am not sure about the spikes down when the heater kicks on - perhaps noise caused by the heater which is a very large load at 5500W.  But perhaps they are suspect as well.  There have been some instances where it appears that the sensors disappear from the system - I have this running on a cron job and the cron job occasionally fails.  I will try to log when this happens.

How do I access the Tasmota logs?  Is this something that I have to set up in Configuration/Configure Logging?

**Reply by barbudor** [reply]
Recent Tasmota logs are visible in the web console
For long history you can use a syslog receiver in your back end and configure Tasmota to send the logs there
Serial logs are also possible but not convenient in your case

**Reply by sfromis** [reply]
With 5500W, you are well above the 20A max load for the Sonoff THR320D.

**Reply by wayner9** [reply]
The THR320D is just a temperature monitor here, it is not doing any switching.  The switching is being done by a POWR3 which has a listed capacity of 5500W.  But I am either misunderstanding something or the POWR3 is reporting energy usage that is way too low.  The heater is running at power of 5500W roughly twice per day for 30 minutes each.  So the usage should be about 5.5kWh.  But Tasmota is only reporting about a quarter of that in the daily usage.  Prior to installing Tasmota the switch was showing usage of about 5-6 kWh per day on eWeLink.  

I am not at home right now so I don't have access to all of the data, but I will start a new thread on the energy reporting issues of the POWR3 later today.

**Reply by sfromis** [reply]
Did you make sure to do the Tasmota calibration of the PowR3? If not, results are dubious.

**Reply by wayner9** [reply]
No, I haven't got around to doing the calibration yet.  But my guess is that the energy reported is roughly 1/4 of actual consumption, but that the power is approximately correct.  Can energy be be off by that much?  And if the power is approximately correct, then how is the energy wrong?  Like if Tasmota is reporting 5500W for 30 minutes then shouldn't Tasmota report that energy consumed is 5500 *30/60 or 2750 kWh?

**Reply by sfromis** [reply]
Without good calibration, values can be worse than inexact. A lot depends on the power factor. I'd simply not assume reasonable numbers without calibration, and not speculate on just how bad it may be.

Assuming that the tub heater is a simple resistive load, and the rating of 5500W is not way off, it might work to run the calibration (while it is heating) using mains voltage, 5500W and calculate the current, like around 23900 mA.
